pkg/controller/clone-controller.go

```diff
@@ -102,162 +185,150 @@ func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
-		// A missing volume would most likely indicate a pod that has been created manually, but also incorrectly defined.
-		return errors.Errorf("Pod does not contain volume %q", ImagePathName)
+		return err
 	}
-
-	glog.V(Vdebug).Infof("processPodItem: getting pvc for key %q", pvcKey)
-	pvc, err := c.pvcFromKey(pvcKey)
-	if err != nil {
-		return errors.WithMessage(err, "could not retrieve pvc from cache")
+	if pvc == nil {
+		return nil
 	}
-	// see if pvc's pod phase anno needs to be added/updated. The update is done only on the target PVC
-	phase := string(pod.Status.Phase)
-	_, exists := pvc.ObjectMeta.Annotations[AnnCloneRequest]
-	if !checkIfAnnoExists(pvc, AnnPodPhase, phase) && exists {
-		pvc, err = setPVCAnnotation(c.clientset, pvc, AnnPodPhase, phase)
-		if err != nil {
-			return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnPodPhase, phase, pvc.Name))
-
-		}
-		glog.V(Vdebug).Infof("processPodItem: pod phase %q annotated in pvc %q", pod.Status.Phase, pvcKey)
-		if phase == "Succeeded" {
-			pvc, err = setPVCAnnotation(c.clientset, pvc, AnnCloneOf, "true")
-			if err != nil {
-				return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnCloneOf, "true", pvc.Name))
-			}
-			glog.V(Vdebug).Infof("processPodItem: CloneOf annotatated in pvc %q", pvcKey)
-		}
+	// filter pvc and decide if the cloning pods should be created
+	if !checkClonePVC(pvc) {
+		return nil
 	}
-	return nil
+	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
+	return c.processPvcItem(pvc)
 }
 
 // Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
 // We forget the key unless `processPvcItem` returns an error in which case the key can be
 // retried.
 func (c *CloneController) ProcessNextPvcItem() bool {
-	key, shutdown := c.pvcQueue.Get()
+	key, shutdown := c.queue.Get()
 	if shutdown {
 		return false
 	}
-	defer c.pvcQueue.Done(key)
+	defer c.queue.Done(key)
 
-	pvc, err := c.pvcFromKey(key)
-	if err != nil || pvc == nil {
-		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: error converting key %v to pvc: %v", key, err))
-	}
-	// filter pvc and decide if the source and target pods should be created
-	if continue_processing, _, _ := checkClonePVC(c.clientset, pvc, false); !continue_processing {
-		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: skipping pvc %q\n", key))
-	}
-	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
-	err = c.processPvcItem(pvc)
+	err := c.syncPvc(key.(string))
 	if err != nil { // processPvcItem errors may not have been logged so log here
 		glog.Errorf("error processing pvc %q: %v", key, err)
 		return true
 	}
 	return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
 }
 
+func (c *CloneController) findClonePodsFromCache(pvc *v1.PersistentVolumeClaim) ([]*v1.Pod, error) {
```

Comment:
why slice return vs. (*srcPod, *tgtPod)?

Reply:
this function isn't the same as the import controller. I don't think they can be shared even unless more refectoring is done. I wouldn't even worry about trying to share them though.

The import controller is looking for a single pod from the cache. Any pod list that is >1 is an error condition.

The clone controller is looking for two pods in the cache. But the error conditions are much more complex. multiple target pods are an error, multiple src pods are an error, pods length > 2 is an error. Success is only when pods <= 2 and there are no duplicate types. 

I'd prefer to see *srcPod, *tgtPod returned because that makes the logic for the work queue that consumes the find function much more straight forward... You wouldn't need the pod iteration loop for example.

Reply:
I thought that the meaning of expectations is to avoid creation of multiple pods (including src and target). Let's talk more about expectations. Will send an invite.

Reply:
expectations does prevent multiple duplicate pods from being created.

Look at the needsSync boolean.  you set an expecation, then you prevent acting on the PVC until the expectation is met. This prevents multiple pods from being created because it forces the controller to observe a Pods creation in the cache before allowing more pods to be created in any capacity.